Add Store Locator

mattysweeps opened the pull request:
Store Locator shows stores on a map. It uses Google Maps for the web page and Google Spanner for the data store.

Commit:
Add store-locator

josephlewis42 commented on storelocator/.gitignore:
```diff
@@ -0,0 +1,4 @@
+.envrc
+
+.mvn
+target
```
I think I'm leaning towards using Spring's recommended `.gitignore` file to keep external contributors from committing their stuff. Thoughts?

```
/target/
!.mvn/wrapper/maven-wrapper.jar
*~

### STS ###
.apt_generated
.classpath
.factorypath
.project
.settings
.springBeans
.sts4-cache

### IntelliJ IDEA ###
.idea
*.iws
*.iml
*.ipr

### NetBeans ###
/nbproject/private/
/build/
/nbbuild/
/dist/
/nbdist/
/.nb-gradle/
```

josephlewis42 commented on storelocator/README.md:
```diff
@@ -0,0 +1,36 @@
+# Setup
+## Credentials
+ * Get an API key for the Google Maps API
+ * Get a Service Account JSON key for a Spanner instance
```
I don't think the user should be creating these. Ideally they'll be able to just use the service broker.

josephlewis42 commented on storelocator/README.md:
```diff
@@ -0,0 +1,36 @@
+ 
+# Technologies Showcased
+* [Google Spanner](https://cloud.google.com/spanner/)
+* [Google Maps Javascript API](https://developers.google.com/maps/documentation/javascript/tutorial)
```
nit: JavaScript

josephlewis42 commented on storelocator/manifest.yml:
```diff
@@ -0,0 +1,3 @@
+---
+applications:
+- name: storelocator
```
Consider including an `env` section with the environment variables you need.

josephlewis42 commented on storelocator/scripts/load-sample-stores.py:
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python
```
I think we should try to put this in the Java app, when the app deploys it could set up the Spanner table and maybe use the creation/deletion of a store as a `/health` check. That way the app is self-contained.

Ideally, a developer will never get to even see an account credential or use the `gcloud` CLI, so it would be handy to show it as a use-case too.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/DatabaseInitializer.java:
```diff
@@ -0,0 +1,84 @@
+  private final StoreService storeService;
+
+  @Autowired
+  DatabaseInitializer(
```
Autowired will fail if the objects are `null`, so I think you can omit this constructor entirely if you just wanted to use the `@Autowired` annotations.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/StoreLocatorProperties.java:
```diff
@@ -0,0 +1,114 @@
+    this.googleResults = googleResults;
+  }
+
+  public static class Map {
```
Would you mind pulling these inner-classes out into a package for clarity? And maybe give Map a different name, like GoogleMap to avoid confusion with java.util.Map

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/StoreLocatorProperties.java:
```diff
@@ -0,0 +1,114 @@
+
+    private double lat;
+    private double lng;
+    private int zoom;
```
Maybe leave a comment about what zoom is and acceptable domain/range of values?

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/StoreLocatorProperties.java:
```diff
@@ -0,0 +1,114 @@
+
+  public static class GoogleResults {
+
+    private boolean enable;
```
I think this should be enabled rather than enable so it doesn't get confused for a method.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/config/CloudFoundryDeploymentCondition.java:
```diff
@@ -0,0 +1,41 @@
+ *
+ * Use as the target class on {@link Conditional} configurations.
+ */
+public class CloudFoundryDeploymentCondition implements Condition {
```
I don't think we need this class or the credentials provider, because we should be able to fish out what we need from the spring boot starter for spanner

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/data/Locatable.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * A physical location given by coordinates.
+ */
+public interface Locatable {
```
I think it's worth dumping this interface for YANGI reasons, the user can always go back and extend or add different endpoints for different types of things.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/data/LocationBounds.java:
```diff
@@ -0,0 +1,100 @@
+      @JsonProperty("lng1") double lng1,
+      @JsonProperty("lat2") double lat2,
+      @JsonProperty("lng2") double lng2) {
+    this.lat1 = Math.min(lat1, lat2);
```
I think this logic fails when you wrap around the globe.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/data/LocationBounds.java:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Since you have the Apache Commons library, I recommend `EqualsBuilder.reflectionEquals(this, o);`, `HashCodeBuilder.reflectionHashCode(this)` and `ToStringBuilder.reflectionToString(this);` to improve extensibility.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/data/Store.java:
```diff
@@ -0,0 +1,188 @@
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
Same here for the equals/toString, etc.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/service/AbstractSpannerLocationService.java:
```diff
@@ -0,0 +1,85 @@
+ */
+public abstract class AbstractSpannerLocationService<S extends Locatable> implements LocationService<S> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(StoreRepository.class);
```
I think StoreRepository.class is a copy/paste error.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/service/AbstractSpannerLocationService.java:
```diff
@@ -0,0 +1,85 @@
+  public void initializeTable(SpannerDatabaseAdminTemplate spannerDatabaseAdminTemplate,
+      SpannerSchemaUtils spannerSchemaUtils) {
+    if (!spannerDatabaseAdminTemplate.tableExists(tableName)) {
+      LOGGER.info("Creating database:table " + spannerDatabaseAdminTemplate.getDatabase() + ":"
```
recommend: LOGGER.info("Creating database: table {}:{}", spannerDatabaseAdminTemplate.getDatabase(), tableName)

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/web/LocationController.java:
```diff
@@ -0,0 +1,63 @@
+@RequestMapping("/api/v1/locations")
+public class LocationController {
+
+  private final Collection<LocationService<?>> repositories;
```
I'm not sure being able to do this all in `/locations` is better than breaking out the endpoints per-table like `/api/v1/locations/stores` because it causes a lot of thought overhead with generics.

I think if all the Locatables come with their own type names, then you're indicating you want to distinguish the different types and they should have different endpoints. Maybe with a single controller that would do the queries in parallel and reduce server-side before returning to the client like this in addition to controllers per-type.

josephlewis42 commented on storelocator/src/main/java/com/google/cloud/servicebroker/samples/storelocator/web/LocationController.java:
```diff
@@ -0,0 +1,63 @@
+   */
+  @GetMapping(produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+  public List<Locatable> locations(LocationBounds bounds) {
+    return repositories.stream()
```
It could be worth doing parallelStream() here because the computations are going to be I/O bound.

josephlewis42 commented on storelocator/src/main/resources/application.properties:
```diff
@@ -0,0 +1,31 @@
+spring.cloud.gcp.spanner.database=${STORE_LOCATOR_SPANNER_DATABASE}
+store-locator.map.key=${STORE_LOCATOR_MAP_KEY}
+# Map configuration
+store-locator.map.lat=47.6
```
Should these be settable, determined from the user's location or maybe the midpoint of all the lat/lons?

josephlewis42 commented on storelocator/src/main/resources/static/js/main.js:
```diff
@@ -0,0 +1,147 @@
+}
+
+// Updates the Vue data
+function updateView(results) {
```
this should probably be in a computed or updated block on the Vue.js object.

josephlewis42 commented on storelocator/src/main/resources/templates/index.html:
```diff
@@ -0,0 +1,78 @@
+
+<script src="/webjars/vue/vue.js"></script>
+<script src="/webjars/jquery/jquery.min.js"></script>
+<script src="https://maps.googleapis.com/maps/api/js?key=NO_KEY&callback=initMap"
```
I would probably load this and JQuery above Vue.js so you can ensure they're initialized once vue starts and won't need the callback anymore.

josephlewis42 reviewed (changes requested):
Overall good, I think there are a few places we can reduce complexity to make this a better jumping-off point for other developers who might want to start with it and go other directions.

Commit:
Add checkstyle to storelocator

Commit:
Some storelocator changes

* CloudFoundry config files deleted; they are not neccesary as spring
starter packages handle scanning VCAP_SERVICES already.
* Moved Properties classes into Config
* Ripped out unused Google Search Results functionality
* Use Apache commons reflection-based methods for equals, hashcode, and toString
* Minor formatting/typo fixes

Commit:
Remove storelocator's Locateable abstraction

The Locateable interface only has 1 implementation: Store.
This commit removes it, which simplifies the code path:

* Store no longer implements Locateable
* No need for a LocationService
* No need for an AbstractSpannerLocationService
* LocationController -> StoreController

Commit:
Remove storelocator's test dependency android-json.

android-json is a transitive dependency of spring-boot-starter-test

Since the project already has a dependency on Google JSON, this resulted
in two of the same package existing on the test classpath.

Commit:
Add storelocator's test data to the app itself.

Note that Spanner doesn't seem to support the `schema.sql` and
`data.sql` approach, so an ApplicationRunner bean is used instead.

The initializeDatabase method was moved from the StoreService.

Also, there is a bug in spring-cloud-gcp-data-spanner, such that the
database creation incorrectly chooses the INT64 type for primitive
doubles. Hence the Store's latitude and longitude fields were updated to
Doubles.

Commit:
Update storelocator's gitignore

Commit:
Small storelocator changes

* Add more comments to application.properties
* Clean up readme so it makes more sense for ClouD Foundry deployments
